### PR TITLE
arch: inline up_switch_context,in arm arm64

### PR DESCRIPTION
**Pull request by hujun260**
## Summary
inline up_switch_context,in arm arm64
when a context switch occurs, up_switch_context is executed. In order to reduce the time taken for context switching, we inline the up_switch_context function.

## Impact
arm,arm64

## Testing
ostest

### CODE REVIEW

File: arch/arm/include/irq.h [hunk 1]
```diff
@@ -76,6 +77,17 @@
 
 #ifndef __ASSEMBLY__
 
+#ifndef up_switch_context
+#define up_switch_context(tcb, rtcb)                              \
```
Comment (anchao) [inline arch/arm/include/irq.h:81]: why not use inline function?

File: arch/arm/src/common/arm_switchcontext.c [hunk 2]
```diff
@@ -1,91 +0,0 @@
...
-
-  nxsched_suspend_scheduler(rtcb);
-
-  /* Are we in an interrupt handler? */
-
-  if (up_interrupt_context())
-    {
-      /* Update scheduler parameters */
-
-      nxsched_resume_scheduler(tcb);
```
Comment (anchao) [inline arch/arm/src/common/arm_switchcontext.c:68]: how about resume/suspend trace?

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



**No, this PR does not meet the NuttX Requirements.**

**Missing Information:**

* **Summary:**
    *  Lacks a clear explanation of *why* inlining `up_switch_context` is necessary. What specific performance benefits are expected? 
    *  Doesn't specify which functional part of the code is being modified (e.g., scheduler, context switching mechanism).
* **Impact:**
    *  **Impact on user:**  While you mention "arm, arm64", it's unclear if users need to make any code changes or configuration adjustments. 
    *  **Impact on build, hardware, documentation, security, compatibility:** All these sections are marked as "arm, arm64" which doesn't provide any meaningful information. You need to state "YES" or "NO" and provide descriptions if applicable.
* **Testing:**
    *  **Build Host(s):**  Specify the operating system, CPU architecture, and compiler used for building NuttX.
    *  **Target(s):**  Be specific about the architecture (e.g., ARM Cortex-M4) and the board configuration used for testing.
    *  **Testing logs:** Provide actual testing logs before and after the change. "ostest" is not sufficient. Include relevant log output that demonstrates the performance improvement or any issues.

**Recommendations:**

1. **Explain the Rationale:**  Clearly articulate the reasons for inlining `up_switch_context`.  Provide specific benchmarks or performance data to justify the change.
2. **Detailed Impact Assessment:** Analyze and describe the potential impact on all aspects mentioned (build, hardware, etc.). If there's no impact, state "NO" explicitly.
3. **Thorough Testing:** Provide detailed testing information, including build host and target specifications, and relevant log snippets that showcase the change's effects. 

By addressing these points, your PR will be better aligned with NuttX's requirements and easier for reviewers to evaluate.

File: arch/arm/include/irq.h (hunk 1 above)
Comment (hujun260) [reply arch/arm/include/irq.h:81]: If we use inline, we need to see the definition of struct tcb_s, but irq.h can't include sched/sched.h

File: arch/arm/src/common/arm_switchcontext.c (hunk 2 above)
Comment (hujun260) [reply arch/arm/src/common/arm_switchcontext.c:68]: I added a commit for a dependency

**Comment by acassis** [comment]
@hujun260 modifications in the context switch requires deep testing, only ostest (you didn't say which arch was used) is not enough. History shows it always introduces regression.

**Comment by acassis** [comment]
Please remove reference to VELAPLATFO-38287

**Comment by hujun260** [comment]
> @hujun260 modifications in the context switch requires deep testing, only ostest (you didn't say which arch was used) is not enough. History shows it always introduces regression.

I understand your concerns. We've tested this PR on multiple products internally.

File: arch/arm64/include/irq.h [hunk 3]
```diff
@@ -423,6 +424,15 @@ static inline_function void up_set_current_regs(uint64_t *regs)
   __asm__ volatile ("msr " "tpidr_el1" ", %0" : : "r" (regs));
 }
 
+#define up_switch_context(tcb, rtcb)                              \
```
Comment (xiaoxiang781216) [inline arch/arm64/include/irq.h:427]: let's create a new patch to inline other arch too

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**